src/Products/Products.Core/ValueObjects/AllergenCollection.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using IGroceryStore.Products.Entities;
+
+namespace IGroceryStore.Products.ValueObjects
+{
+    internal sealed class AllergenCollection : IEnumerable<Allergen>, IEquatable<AllergenCollection>
+    {
+        private ISet<Allergen> _allergens;
+        public static AllergenCollection Empty => new(Enumerable.Empty<Allergen>());
+
+        public IReadOnlySet<Allergen> Allergens
+        {
+            get { return (IReadOnlySet<Allergen>)_allergens; }
+            private set { _allergens = (ISet<Allergen>)value; }
+        }
+
+        public AllergenCollection(IEnumerable<Allergen> allergens)
+        {
+            _allergens = allergens.ToHashSet();
+        }
+
+        public bool Equals(AllergenCollection? other)
+        {
+            if (other == null) return false;
+
+            return _allergens.SetEquals(other!._allergens);
+        }
+
+        public override bool Equals(object? other)
+        {
+            return Equals(other as AllergenCollection);
+        }
+        public IEnumerator<Allergen> GetEnumerator()
+        {
+            return _allergens.GetEnumerator();
+        }
+        public override int GetHashCode()
+        {
+            return _allergens.Select(x => x != null ? x.GetHashCode() : 0).Aggregate((x, y) => x ^ y);
```

Comment:
use HashCode class and add cache for hash code 
```suggestion
           var hash = new HashCode();
            foreach (var allergen in _allergens)
            {
                hash.Add(allergen);
            }
            return hash.ToHashCode();
```

Reply:
I feel bad about that caching. Adding allergen should change hashcode and i think i don't want to recalculate hash every time i make change on that set. Maybe I should add flag or something? What do you think?